<a href="https://colab.research.google.com/github/Kate5-7-2021/Kate_INFO5731_Spring2021/blob/main/In_class_exercise_08_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **The eighth in-class-exercise (20 points in total, 3/30/2021)**

The data for this exercise is from the dataset you created from assignment three. Please perform answer the following questions based on your data:

## (1) (10 points) Write a python program to extract the sentiment related terms from the corpus. You may use python package such as polyglot or external lexicon resources in the question. Rank the sentiment related terms by frequency.

In [4]:
import csv        # I redue the assignment 2 & 3 to get the new dataset because I think in the previous dataset all my data were repeated
import pandas as pd     # This is a new link for my new csv file
                        # https://github.com/Kate5-7-2021/Kate_INFO5731_Spring2021/blob/main/results-final.csv
df = pd.read_csv('/content/results-final.csv')
df


document_id                                         clean_text sentiment
0              0  bwho drtedro resignnl dec wechat group publish...   neutral
1              1  bsenatorhagerti wuhancoronaviru spread world t...  positive
2              2  bjuliensorel chainhomotopi thelogankyl busi xf...   neutral
3              3  bdrmadej i encourag everyon get vaccin necessa...  positive
4              4  bchainhomotopi juliensorel thelogankyl busi xf...   neutral
..           ...                                                ...       ...
123          123  bswami you show understand civiliz issu herenn...   neutral
124          124  bglobaltimesnew thatxexx want put ccp dog cage...  negative
125          125  btypediabet final ad massachusettsxexx list el...  positive
126          126  btelglobalhealth sneweyi good luck california ...  negative
127          127  buk u criticis who covid report accus china wi...  negative

[128 rows x 3 columns]

In [10]:
import nltk
#nltk.download('stopwords')
from nltk.corpus import stopwords

In [12]:
!pip install polyglot

     |████████████████████████████████| 133kB 10.1MB/s 
  Created wheel for polyglot: filename=polyglot-16.7.4-py2.py3-none-any.whl size=52557 sha256=e51aceea7860042f0a85461909e1befa42b82215da628b32766361f12f244400
  Stored in directory: /root/.cache/pip/wheels/5e/91/ef/f1369fdc1203b0a9347d4b24f149b83a305f39ab047986d9da
Successfully built polyglot


In [14]:
!pip install pyicu

     |████████████████████████████████| 235kB 7.4MB/s 
  Created wheel for pyicu: filename=PyICU-2.6-cp37-cp37m-linux_x86_64.whl size=1306402 sha256=071356cdaac519b366a7953b70afb26b8f3e7fe5efb5ce8a2ba76213ddba2d8f
  Stored in directory: /root/.cache/pip/wheels/31/21/2f/1c91831e8a93537ab21f6b4b935781b681104635fdb0315791
Successfully built pyicu


In [16]:
! pip install pycld2

     |████████████████████████████████| 41.4MB 104kB/s 
  Created wheel for pycld2: filename=pycld2-0.41-cp37-cp37m-linux_x86_64.whl size=9834263 sha256=78e2067423c845fb96336f4f621e3c3c5a9133cf025a8d77548b8ee236401104
  Stored in directory: /root/.cache/pip/wheels/c6/8f/e9/08a1a8932a490175bd140206cd86a3dbcfc70498100de11079
Successfully built pycld2


In [17]:
from polyglot.downloader import downloader             #this code shows all languages that polyglot has polarity lexicons for them
print(downloader.supported_languages_table("sentiment2", 3))

  1. Kapampangan                2. Italian                    3. Upper Sorbian            
  4. Sakha                      5. Hindi                      6. French                   
  7. Spanish; Castilian         8. Vietnamese                 9. Arabic                   
 10. Macedonian                11. Pashto, Pushto            12. Bosnian-Croatian-Serbian 
 13. Egyptian Arabic           14. Norwegian Nynorsk         15. Sundanese                
 16. Sicilian                  17. Azerbaijani               18. Bulgarian                
 19. Yoruba                    20. Tajik                     21. Georgian                 
 22. Tatar                     23. Galician                  24. Malagasy                 
 25. Uighur, Uyghur            26. Amharic                   27. Venetian                 
 28. Yiddish                   29. Norwegian                 30. Alemannic                
 31. Estonian                  32. West Flemish              33. Divehi; Dhivehi; Mald... 

In [49]:
! pip install morfessor

In [50]:
from polyglot.text import Text

text = Text("bswami you show understand civiliz issu herenn")

In [51]:
%%bash
polyglot download sentiment2.en

[polyglot_data] Downloading package sentiment2.en to
[polyglot_data]     /root/polyglot_data...
[polyglot_data]   Package sentiment2.en is already up-to-date!


In [52]:
print("{:<16}{}".format("Word", "Polarity")+"\n"+"-"*30)      #to test the polarity of the text
for w in text.words:
    print("{:<16}{:>2}".format(w, w.polarity))

Word            Polarity
------------------------------
bswami           0
you              0
show             0
understand       0
civiliz          0
issu             0
herenn           0


In [99]:
from textblob import TextBlob
df['clean_text'] = df['clean_text'].apply(lambda x: str(TextBlob(x).correct()))

In [90]:
df.sentiment.value_counts()

positive    54
negative    50
neutral     23
neagtive     1
Name: sentiment, dtype: int64

In [105]:
from polyglot.text import Text
s_words = []
for line in df['clean_text']:
  text = Text(line)
  for w in text.words:
      s_words.append(w)
print(s_words)


Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to detect the language reliably.
Detector is not able to dete

['who', 'drtedro', 'resigned', 'dec', 'what', 'group', 'publish', 'present', 'sarslik', 'virusnjan', 'china', 'tell', 'whoxexxa', 'httpstcoaacuuwjmk', 'bsenatorhagerti', 'wuhancoronaviru', 'spread', 'world', 'thank', 'who', 'help', 'cap', 'cap', 'concealxexxa', 'httpstcoyphehfn', 'bjuliensorel', 'chainhomotopi', 'thelogankyl', 'busy', 'xfxfxx', 'due', 'sad', 'i', 'told', 'one', 'hashtag', 'wuhancoronaviru', 'bdrmadej', 'i', 'encourage', 'everyone', 'get', 'vaccine', 'necessary', 'consider', 'care', 'potentixexxa', 'httpstcowzlpvsa', 'bchainhomotopi', 'juliensorel', 'thelogankyl', 'busy', 'xfxfxx', 'seem', 'guy', 'donxexxt', 'get', 'i', 'told', 'guy', 'alreadyxexxa', 'httpstcowlkphh', 'the', 'end', 'week', 'reduce', 'burden', 'alley', 'virus', 'american', 'hospital', 'week', 'overduexexxa', 'httpstcouyvobizhh', 'bdelglenndavi', 'the', 'two', 'week', 'suppose', 'end', 'one', 'year', 'ago', 'now', 'told', 'ostracizedxexxa', 'httpstcokyiaqtsxh', 'bglobaltimesnew', 'xexxctechn', 'testsxexxd

In [106]:
from collections import Counter
freq = Counter(s_words)
df1 = pd.DataFrame(list(freq.most_common()), columns=['word', 'frequency'])
df1.index = list(range(1, len(freq.most_common())+1))
df1


word  frequency
1                cap         53
2    wuhancoronaviru         45
3                who         42
4                 he         41
5            freedom         39
..               ...        ...
527             back          1
528        classroom          1
529             like          1
530      workinxexxa          1
531  httpstconfazueq          1

[531 rows x 2 columns]

#(2) (10 points) Compare the performance of the following tools in sentiment identification: TextBlob (https://textblob.readthedocs.io/en/dev/), VADER (https://github.com/cjhutto/vaderSentiment), TFIDF-based Support Vector Machine (SVM) (Split your data into training and testing data). Take your own annotation as the standard answers.
Reference code: https://towardsdatascience.com/fine-grained-sentiment-analysis-in-python-part-1-2697bb111ed4


##**TextBlob Sentiment Identification**

In [233]:
for word in df['sentiment']:
  
  if word == 'positive':
    encoded = 1
  elif word == 'negative':
    encoded = -1
  elif word == 'neutral':
    encode = 0

s_real = []
for word in df['sentiment']:
   s_real.append(Encoder(word))


In [234]:
from textblob import TextBlob
tb_s = []
for line in df['clean_text']:
  polarity = TextBlob(line).sentiment.polarity
  if polarity > 0:
    tb_s.append(1)
  elif polarity < 0:
    tb_s.append(-1)
  elif polarity == 0.0:
    tb_s.append(0)
df2 = pd.DataFrame(list(zip(df['clean_text'], s_real, tb_s)), columns = ['clean_text', 'manual_polarity', 'tb_polarity'])
df2


clean_text  ...  tb_polarity
0    who drtedro resigned dec what group publish pr...  ...            0
1    bsenatorhagerti wuhancoronaviru spread world t...  ...            0
2    bjuliensorel chainhomotopi thelogankyl busy xf...  ...           -1
3    bdrmadej i encourage everyone get vaccine nece...  ...            0
4    bchainhomotopi juliensorel thelogankyl busy xf...  ...            1
..                                                 ...  ...          ...
123  swamp you show understand civil issue herennth...  ...            0
124  bglobaltimesnew thatxexx want put cap dog cage...  ...            0
125  btypediabet final ad massachusettsxexx list fi...  ...            0
126  btelglobalhealth sneweyi good luck california ...  ...            1
127  but u criticism who could report accuse china ...  ...            0

[128 rows x 3 columns]

In [307]:
from sklearn.metrics import f1_score, accuracy_score
tb_accuracy = accuracy_score(df2['manual_polarity'], df2['tb_polarity'])*100
tb_score = f1_score(df2['manual_polarity'], df2['tb_polarity'], average = 'macro')
print("Accuracy of TextBlob is {0} and F1 Score is {1}".format(tb_accuracy, tb_score))


Accuracy of TextBlob is 43.75 and F1 Score is 0.4251563251563251


##**Vader Sentiment Identification**

In [236]:
import nltk
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [238]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer
va = SentimentIntensityAnalyzer()
va_s = []
for line in df['clean_text']:
  polarity = va.polarity_scores(line)['compound']
  if polarity > 0:
    va_s.append(1)
  elif polarity < 0:
    va_s.append(-1)
  elif polarity == 0.0:
    va_s.append(0)
df3 = pd.DataFrame(list(zip(df['clean_text'], s_real , va_s)), columns = ['clean_text', 'manual_polarity', 'va_polarity'])
df3


clean_text  ...  va_polarity
0    who drtedro resigned dec what group publish pr...  ...           -1
1    bsenatorhagerti wuhancoronaviru spread world t...  ...            1
2    bjuliensorel chainhomotopi thelogankyl busy xf...  ...           -1
3    bdrmadej i encourage everyone get vaccine nece...  ...            1
4    bchainhomotopi juliensorel thelogankyl busy xf...  ...            0
..                                                 ...  ...          ...
123  swamp you show understand civil issue herennth...  ...            0
124  bglobaltimesnew thatxexx want put cap dog cage...  ...           -1
125  btypediabet final ad massachusettsxexx list fi...  ...            0
126  btelglobalhealth sneweyi good luck california ...  ...            1
127  but u criticism who could report accuse china ...  ...           -1

[128 rows x 3 columns]

In [239]:
from sklearn.metrics import f1_score, accuracy_score
va_accuracy = accuracy_score(df3['manual_polarity'], df3['va_polarity'])*100
va_f1_score = f1_score(df3['manual_polarity'], df3['va_polarity'], average = 'macro' )
print("Accuracy of vader is {0} and F1 Score is {1}".format(va_accuracy, va_f1_score))


Accuracy of vader is 61.71875 and F1 Score is 0.5864970313825276


##**SVM Sentiment Identification**

In [329]:
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer


In [337]:
TFid_vectorizer = TfidfVectorizer(min_df = 5, max_df = 0.8, sublinear_tf = True, use_idf = True)

Train_X, Test_X, Train_Y, Test_Y = train_test_split(df['clean_text'],df['sentiment'],test_size=0.33)
Encode = LabelEncoder()
Train_Y = Encode.fit_transform(Train_Y)
Test_Y = Encode.fit_transform(Test_Y)
Train_X_vec = TFid_vectorizer.fit_transform(Train_X)
Test_X_vec = TFid_vectorizer.transform(Test_X)



In [338]:
from sklearn import svm
from sklearn.metrics import classification_report

svm_model = svm.SVC(kernel='linear')
svm_model.fit(Train_X_vec, Train_Y)
predicted = svm_model.predict(Test_X_vec)
output = classification_report(Test_Y, predicted, output_dict=True)
output

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


{'0': {'f1-score': 0.0, 'precision': 0.0, 'recall': 0.0, 'support': 17},
 '1': {'f1-score': 0.3783783783783784,
  'precision': 0.25,
  'recall': 0.7777777777777778,
  'support': 9},
 '2': {'f1-score': 0.0, 'precision': 0.0, 'recall': 0.0, 'support': 17},
 '3': {'f1-score': 0.0, 'precision': 0.0, 'recall': 0.0, 'support': 0},
 'accuracy': 0.16279069767441862,
 'macro avg': {'f1-score': 0.0945945945945946,
  'precision': 0.0625,
  'recall': 0.19444444444444445,
  'support': 43},
 'weighted avg': {'f1-score': 0.07919547454431176,
  'precision': 0.05232558139534884,
  'recall': 0.16279069767441862,
  'support': 43}}

##**Results**
#### TextBlob: Accuracy of TextBlob is 43.75 and F1 Score is 0.4251563251563251
                                        
#### Vader: Accuracy of vader is 61.71875 and F1 Score is 0.5864970313825276 

#### SVM: Based on the above results of the SVM identification tools the value of the precision and recall both present and after each running time either by changing the min_df, max_df (for example changing the min_df to 4 or 6 and changing the max_df to 0.7 or 0.9) one can get different results for F1 score, min_df, max_df. However, in each try I got the F1 score (which is the sign of getting false positive and false negative) lower than two other tools. The great value for F1 score is 1 and between the other tools (TextBlob and Vader), Vader with higher accuracy and higher F1 score seems to have better performance for my sentiment identification.    


